csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -254,8 +270,17 @@ newtype TTranslatedElement =
     )
   } or
   // A local declaration
-  TTranslatedDeclaration(LocalVariableDeclExpr entry)
-
+  TTranslatedDeclaration(LocalVariableDeclExpr entry) {
+    // foreach var decl and init is treated separately,
+    // because foreach needs desugaring
+    not entry.getParent() instanceof ForeachStmt
+  } or
+  // A compiler generated element, generated by `generatedBy` during the 
+  // desugaring process
+  TTranslatedCompilerGeneratedElement(Element generatedBy, int index) {
+    canCreateCompilerGeneratedElement(index)
```

Comment:
I've modified `canCreateCompilerGeneratedElement`, now we create 15 copies only for the elements that need desugaring.

```
private predicate canCreateCompilerGeneratedElement (Element generatedBy, int nth) {
  (
    generatedBy instanceof ForeachStmt or
    generatedBy instanceof LockStmt or
    generatedBy instanceof DelegateCreation or
    generatedBy instanceof DelegateCall
  ) and
  // For now we allow a max of 15 compiler generated elements
  nth in [0 .. 14]
}
```

Reply:
That is certainly a lot better. I wonder, though, if it would be possible to make the IPA type only contain exactly the copies that are needed? This usually means making the IPA type mutually recursive with a generating predicate.